sklearn/metrics/tests/test_classification.py

```diff
@@ -724,9 +725,39 @@ def test_cohen_kappa():
     )
 
 
-def test_matthews_corrcoef_nan():
-    assert matthews_corrcoef([0], [1]) == 0.0
-    assert matthews_corrcoef([0, 0], [0, 1]) == 0.0
+@pytest.mark.parametrize("zero_division", [0, 1, np.nan])
+@pytest.mark.parametrize(
+    "y_true, y_pred",
+    [
+        ([0], [1]),
+        ([0, 0], [1, 1]),
+        ([], []),
+    ],
+)
+def test_matthews_corrcoef_nan(zero_division, y_true, y_pred):
+    with warnings.catch_warnings(record=True) as record:
+        mcc = matthews_corrcoef(y_true, y_pred, zero_division=zero_division)
+        assert not record
+
+    if np.isnan(zero_division):
+        assert np.isnan(mcc)
+    else:
+        assert mcc == zero_division
+
+
+@pytest.mark.parametrize(
+    "y_true, y_pred",
+    [
+        ([0], [1]),
+        ([0, 0], [1, 1]),
+        ([], []),
+    ],
+)
+def test_matthews_corrcoef_nan_warn(y_true, y_pred):
+    with warnings.catch_warnings(record=True) as record:
+        mcc = matthews_corrcoef(y_true, y_pred, zero_division="warn")
+        assert len(record) == 1
+        assert mcc == 0.0
```

Comment:
What I meant is the following general test:

```python
@pytest.mark.parametrize("zero_division", [0, 1, np.nan])
@pytest.mark.parametrize("y_true, y_pred", [([0], [0]), ([], [])])
@pytest.mark.parametrize("metric", [
    jaccard_score,
    matthews_corrcoef,
    f1_score,
    partial(fbeta_score, beta=1),
    precision_score,
    recall_score,
])
def test_zero_division_nan_no_warning(metric, y_true, y_pred, zero_division):
    """Check the behaviour of `zero_division` when setting to 0, 1 or np.nan.
    No warnings should be raised.
    """
    with warnings.catch_warnings():
        warnings.simplefilter("error")
        result = metric(y_true, y_pred, zero_division=zero_division)

    if np.isnan(zero_division):
        assert np.isnan(result)
    else:
        assert result == zero_division


@pytest.mark.parametrize("y_true, y_pred", [([0], [0]), ([], [])])
@pytest.mark.parametrize("metric", [
    jaccard_score,
    matthews_corrcoef,
    f1_score,
    partial(fbeta_score, beta=1),
    precision_score,
    recall_score,
])
def test_zero_division_nan_warning(metric, y_true, y_pred):
    """Check the behaviour of `zero_division` when setting to "warn".
    A `UndefinedMetricWarning` should be raised.
    """
    with pytest.warns(UndefinedMetricWarning):
        result = metric(y_true, y_pred, zero_division="warn")
    assert result == 0.0
```

Here, I don't include `precision_recall_f1_score_support` and `classification_report` because they are a bit different.

The idea is not to test all potential cases that trigger the warning but the general behaviour when the `UndefinedMetricWarning` is triggered.